machine/xen/xenstore_client.go

```diff
@@ -0,0 +1,194 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2022, Unikraft GmbH and The KraftKit Authors.
+// Licensed under the BSD-3-Clause License (the "License").
+// You may not use this file except in compliance with the License.
+
+//go:build xen
+// +build xen
+
+package xen
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"strings"
+
+	"kraftkit.sh/log"
+	"xenbits.xenproject.org/git-http/xen.git/tools/golang/xenlight"
+)
+
+type XenstoreOperation uint32
+
+const (
+	WatchOp    XenstoreOperation = 4
+	UnwatchOp  XenstoreOperation = 5
+	WatchEvent XenstoreOperation = 15
+	Error      XenstoreOperation = 16
+
+	XenStorePathFmt = "/local/domain/%d"
+)
+
+func XenstoreSocketPath() string {
+	xenstorepath := "/var/run/xenstored"
+	if path := os.Getenv("XENSTORED_PATH"); path != "" {
+		xenstorepath = path
+	}
+
+	return strings.Join([]string{xenstorepath, "socket"}, "/")
+}
+
+type XsHeader struct {
+	Op     XenstoreOperation
+	ReqID  uint32
+	TxID   uint32
+	Lenght uint32
+}
+
+type XsPacket struct {
+	Header XsHeader
+	Data   []byte
+}
+
+type baseWatcher struct {
+	closeSignal chan struct{}
+
+	domID  uint32
+	conn   net.Conn
+	xsPath string
+	token  string
+}
+
+type Watcher interface {
+	Watch(ctx context.Context) (chan struct{}, error)
+	Close()
+}
+
+func NewWatcher(domID xenlight.Domid) (Watcher, error) {
+	conn, err := net.Dial("unix", XenstoreSocketPath())
+	if err != nil {
+		return nil, err
+	}
+
+	return &baseWatcher{
+		domID:       uint32(domID),
+		conn:        conn,
+		xsPath:      fmt.Sprintf(XenStorePathFmt, uint32(domID)),
+		token:       "kraftkit" + fmt.Sprintf("%d", uint32(domID)),
+		closeSignal: make(chan struct{}),
+	}, nil
+}
+
+func (packet *XsPacket) pack() []byte {
+	data := make([]byte, 0)
+	data = binary.LittleEndian.AppendUint32(data, uint32(packet.Header.Op))
+	data = binary.LittleEndian.AppendUint32(data, packet.Header.ReqID)
+	data = binary.LittleEndian.AppendUint32(data, packet.Header.TxID)
+	data = binary.LittleEndian.AppendUint32(data, packet.Header.Lenght)
+
+	data = append(data, packet.Data...)
+
+	return data
+}
+
+func unpack(data []byte) XsPacket {
```

Comment:
I think you'll need a nice ascii comment to explain these 2 functions for future maintenance